lib/AST/RequirementMachine/RewriteSystem.h

```diff
@@ -44,6 +46,11 @@ class Rule final {
   Term LHS;
   Term RHS;
 
+  /// The written requirement ID, which can be used to index into the
+  /// \c WrittenRequirements array in the rewrite system to retrieve
+  /// the structural requirement.
+  Optional<unsigned> requirementID;
```

Comment:
Can you squeeze this into the below bitfield? Use a biased representation where 0 means None and x+1 means Some(x). 16 bits ought to be enough